src/findFocalPlane.cpp

```diff
@@ -0,0 +1,184 @@
+  return x;
+};
+
+TVector3 avgDir(std::vector<TVector3> dvec){
+  double xavg = 0;
+  double yavg = 0;
+  double zavg = 0;
+  double n = 0;
+  for(int i = 0; i < dvec.size(); i++){
+    xavg += dvec[i].X();
+    yavg += dvec[i].Y();
+    zavg += dvec[i].Z();
+    n+=1.0;
+  }
+  
+  TVector3 outdir;
+  if(n > 0) outdir.SetXYZ(xavg/n, yavg/n, zavg/n);
+  return outdir;
+};
+
+int main(int argc, char** argv) {
+  // setup
+  TString infileN="sim_rich_vis.root";
+  if(argc>1) infileN = TString(argv[1]);
+
+  RDataFrame dfIn("events",infileN.Data());
+  TString outfileN = infileN;
+  outfileN(TRegexp("\\.root$"))=".";
+  TFile *outfile = new TFile(outfileN+"plots.root","RECREATE");
+  gStyle->SetOptStat(0);
+
+  /* lambdas
+   * - most of these transform an `RVec<T1>` to an `RVec<T2>` using `VecOps::Map` or `VecOps::Filter`
+   * - see NPdet/src/dd4pod/dd4hep.yaml for POD syntax
+   */
+  cout << 0 << endl;
+  // calculate number of hits
+  auto numHits = [](RVec<SimTrackerHitData> hits) { return hits.size(); };
+  // calculate momentum magnitude for each particle (units=GeV)
+  // TODO: edm4hep::Vector3f really has no magnitude function!?
+  auto momentum = [](RVec<MCParticleData> parts){
+    return Map(parts,[](auto p){
+        auto mom = p.momentum;
+        return sqrt( mom[0]*mom[0] + mom[1]*mom[1] + mom[2]*mom[2] );
+        });
+  };
+  // filter for thrown particles
+  auto isThrown = [](RVec<MCParticleData> parts){
+    return Filter(parts,[](auto p){
+        return p.generatorStatus==1;
+        });
+  };    
+  // get positions for each hit (units=cm)
+  auto hitPos = [](RVec<SimTrackerHitData> hits){ return Map(hits,[](auto h){ return h.position; }); };
+  auto hitPosX = [](RVec<Vector3d> v){ return Map(v,[](auto p){ return p.x; }); };
+  auto hitPosY = [](RVec<Vector3d> v){ return Map(v,[](auto p){ return p.y; }); };
+  auto hitPosZ = [](RVec<Vector3d> v){ return Map(v,[](auto p){ return p.z; }); };
+
+  auto findReflected = [](RVec<MCParticleData> photons){ // want to skip non-reflected photons
+    return Map( photons,[](auto p){
+      auto imom = p.momentum;
+      auto fmom = p.momentumAtEndpoint;
+      return !(imom==fmom);
+    });
+  };
+  auto dirVec = [](RVec<MCParticleData> parts){
+    return Map(parts,[](auto p){
+      auto dMom = TVector3(p.momentumAtEndpoint.x,p.momentumAtEndpoint.y,p.momentumAtEndpoint.z);
+      return dMom.Unit();
+      //return dMom;
+    });
+  };
+  auto endVec = [](RVec<MCParticleData> parts){
+    return Map(parts,[](auto p){
+      return TVector3(p.endpoint.x/10., p.endpoint.y/10., p.endpoint.z/10.);
+    });
+  };        
+  
+  // transformations
+  auto df1 = dfIn
+    .Define("reflectedPhotons",findReflected,{"MCParticles"})
+    .Define("direcVec",dirVec,{"MCParticles"})
+    .Define("endVec",endVec,{"MCParticles"})
+    ;
+  
+  df1.Foreach(
+	      [d,e,fp,dirout,nphotons,nAcc](RVec<bool> refl, RVec<TVector3> dir, RVec<TVector3> end){		
```

Comment:
Running `make` results in warnings on this line, such as
```
src/findFocalPlane.cpp: In function ‘int main(int, char**)’:
src/findFocalPlane.cpp:145:16: warning: capture of variable ‘d’ with non-automatic storage duration
  145 |               [d,e,fp,dirout,nphotons,nAcc](RVec<bool> refl, RVec<TVector3> dir, RVec<TVector3> end){
      |                ^
src/findFocalPlane.cpp:32:23: note: ‘std::vector<TVector3> d’ declared here
   32 | std::vector<TVector3> d, e, fp, dirout;
      |                       ^
```
Did you mean to capture these variables by reference?

Reply:
I did since I needed some way to store the values from N (number of photons per beam) entries in a row, but I'll try to see if there's a better solution that doesn't throw an error.

Reply:
Turns out didn't really have to capture them, silly mistake.

Reply:
Not a silly mistake. Today I learned that a lambda with an empty capture list can still access and mutate globally scoped variables.